### DISCUSSION TITLE
Predicting Node & Edge Features of a Graph

### DISCUSSION THREAD

**Discussion by sac3tf** [danielegrattarola/spektral#437] [General]
Hello,

I am very very new to the GNN space, but have a strong inclination that this space can help solve my problem. I have a physical network of directional nodes and edges that have been mapped out, but in some areas I have missing features about the nodes and networks (both categorical features and continuous). I have been reading through Spektral [documentation](https://graphneural.network/getting-started/) and saw that labels can only be of two types (Node level and graph level). I am wondering if it's possible to utilize Spektral for the task of predicting edge-level labels/features or if it can only be leveraged for node level (I am not interested in graph classification at this time). Or if the preferred method would be to take the edge features and combine them into node features, etc. Any guidance or thoughts here would be greatly appreciated!

Thank you in advance!

**Comment by danielegrattarola** [comment]
>I am wondering if it's possible to utilize Spektral for the task of predicting edge-level labels/features or if it can only be leveraged for node level 

That limitation only applies if you want to use the data loading utilities and the Keras `model.fit` approach. In your case, you just have to write your own training loop as you would in standard TF. Here's a [simple example](https://github.com/danielegrattarola/spektral/blob/master/examples/node_prediction/citation_gcn_custom.py), although you can find better templates for TF in general.

To model edge features you can either compute pairwise features from the nodes (might be expensive depending on how you implement it, I wouldn't recommend it to a novice), or you can use [XENet layers](https://graphneural.network/layers/convolution/#xenetconv) which are currently the only layers in Spektral that support edge representation learning.

A little tinkering might be required here as this is a non-standard use case.

Cheers
